src/compute/src/render/top_k.rs

```diff
@@ -317,6 +332,143 @@ where
             // TODO(#7331): Here we discard the arranged output.
             result.as_collection(|_k, v| v.clone())
         }
+
+        fn render_intra_ts_thinning<G>(
+            collection: Collection<G, Row, Diff>,
+            order_key: Vec<mz_expr::ColumnOrder>,
+            limit: usize,
+        ) -> Collection<G, Row, Diff>
+        where
+            G: Scope,
+            G::Timestamp: Lattice,
+        {
+            let mut aggregates = HashMap::new();
+            let mut vector = Vec::new();
+            collection
+                .inner
+                .unary_notify(
+                    Pipeline,
+                    "TopKIntraTimeThinning",
+                    [],
+                    move |input, output, notificator| {
+                        while let Some((time, data)) = input.next() {
+                            data.swap(&mut vector);
+                            let agg_time = aggregates
+                                .entry(time.time().clone())
+                                .or_insert_with(HashMap::new);
+                            for (row, record_time, diff) in vector.drain(..) {
+                                let monoid = monoids::Top1Monoid {
+                                    row,
+                                    order_key: order_key.clone(),
+                                };
+                                let topk = agg_time.entry(record_time).or_insert_with(move || {
+                                    topk_agg::TopKBatch::new(limit.try_into().expect("must fit"))
+                                });
+                                topk.update(monoid, diff);
+                            }
+                            notificator.notify_at(time.retain());
+                        }
+
+                        // pop completed aggregates, send along whatever
+                        notificator.for_each(|time, _, _| {
+                            if let Some(aggs) = aggregates.remove(time.time()) {
+                                let mut session = output.session(&time);
+                                for (record_time, topk) in aggs {
+                                    session.give_iterator(topk.into_iter().map(
+                                        |(monoid, diff)| (monoid.row, record_time.clone(), diff),
+                                    ))
+                                }
+                            }
+                        });
+                    },
+                )
+                .as_collection()
+        }
+    }
+}
+
+/// Types for in-place intra-ts aggregation of monotonic streams.
+pub mod topk_agg {
+    use smallvec::SmallVec;
+
+    pub struct TopKBatch<T> {
+        updates: SmallVec<[(T, i64); 16]>,
+        clean: usize,
+        limit: i64,
+    }
```

Comment:
This looks a lot like a `ChangeBatch`; is there a specific reason it isn't that (plus the `limit` and a method to prune based on the limit).

Reply:
It's not an accident, I did copy the methods from `ChangeBatch`. The thing I couldn't do with `ChangeBatch` is remove the records that I don't need anymore every time compaction happens. I couldn't think of an API that would make sense to add to `ChangeBatch` to allow for that customization but maybe you have an idea?

Reply:
@frankmcsherry this is the only outstanding comment. I pulled in `differential_dataflow::consolidation` to do the consolidation but that locks us in with the `Ord` trait. I can put the manual consolidation back here and sort based on a shared `Vec<ColumnOrder>`. What do you prefer?

The `Top1` plan suffers from the same problem if that makes you feel any better :sweat_smile:

Reply:
It does not! :D

We should either have an issue open, or a `// TODO`, or a fix. This seems like a fine place to eventually do the work for both cases. You needn't do it here right now, but we should open up an issue if we leave it undone!